overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,373 @@
+)
+
+// DeviceManager is responsible for managing the device identity and device
+// policies.
+type DeviceManager struct {
+	state      *state.State
+	keypairMgr asserts.KeypairManager
+	runner     *state.TaskRunner
+}
+
+// Manager returns a new device manager.
+func Manager(s *state.State) (*DeviceManager, error) {
+	runner := state.NewTaskRunner(s)
+
+	keypairMgr, err := asserts.OpenFSKeypairManager(dirs.SnapDeviceDir)
+	if err != nil {
+		return nil, err
+
+	}
+
+	m := &DeviceManager{state: s, keypairMgr: keypairMgr, runner: runner}
+
+	runner.AddHandler("generate-device-key", m.doGenerateDeviceKey, nil)
+	runner.AddHandler("request-serial", m.doRequestSerial, nil)
+
+	return m, nil
+}
+
+func (m *DeviceManager) ensureOperational() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+	// XXX: auth.Device/SetDevice should probably move to devicestate
+	// they are not quite just about auth and also we risk circular imports
+	// (auth will need to mediate bits from devicestate soon)
+	device, err := auth.Device(m.state)
+	if err != nil {
+		return err
+	}
+
+	if device.Serial != "" {
+		// serial is set, we are all set
+		return nil
+	}
+
+	if device.Brand == "" || device.Model == "" {
+		// need first-boot, loading of model assertion info
+		if release.OnClassic {
+			// TODO: are we going to have model assertions on classic or need will need to cheat here?
+			return nil
+		}
+		// cannot proceed yet, once first boot is done these will be set
+		// and we can pick up from there
+		return nil
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "become-operational" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// XXX: some of these will need to be split and use hooks
+	// retries might need to embrace more than one "task" then,
+	// need to be careful
+
+	genKey := m.state.NewTask("generate-device-key", i18n.G("Generate device key"))
+	requestSerial := m.state.NewTask("request-serial", i18n.G("Request device serial"))
+	requestSerial.WaitFor(genKey)
+
+	chg := m.state.NewChange("become-operational", i18n.G("Setting up device identity"))
+	chg.AddAll(state.NewTaskSet(genKey, requestSerial))
+
+	return nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceManager) Ensure() error {
+	err := m.ensureOperational()
+	if err != nil {
+		return err
+	}
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *DeviceManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *DeviceManager) Stop() {
+	m.runner.Stop()
+}
+
+var (
+	keyLength = 4096
+	// XXX: a 2nd different URL for nonce?
+	// TODO: this will come as config from the gadget snap
+	serialRequestURL = "https://serial.request" // XXX dummy value!
+)
+
+func (m *DeviceManager) doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var keyID string
+	// XXX: make this a field of DeviceState?
```

Comment:
Yes, sounds correct. It's awkward to have some of the device properties under a Device field, and others at the root of the state with a "device-" prefix.